kubernetes/src/main/java/com/linecorp/armeria/client/kubernetes/ArmeriaWebSocket.java

```diff
@@ -0,0 +1,176 @@
+        return true;
+    }
+
+    @Override
+    public long queueSize() {
+        return pending.get();
+    }
+
+    @Override
+    public void request() {
+        // request() is invoked after listener.onOpen() is called.
+        assert inboundSubscription != null;
+        inboundSubscription.request(1);
+    }
+
+    @Override
+    public void onSubscribe(Subscription s) {
+        inboundSubscription = s;
+        request();
+        listener.onOpen(this);
+    }
+
+    @Override
+    public void onNext(WebSocketFrame webSocketFrame) {
+        lastFrameType = webSocketFrame.type();
+        final boolean last = webSocketFrame.isFinalFragment();
+        switch (webSocketFrame.type()) {
+            case TEXT:
+                onText(webSocketFrame.text(), last);
+                break;
+            case BINARY:
+                onBinary(webSocketFrame.array(), last);
+                break;
+            case CONTINUATION:
+                if (lastFrameType == WebSocketFrameType.TEXT) {
+                    onText(webSocketFrame.text(), last);
+                } else if (lastFrameType == WebSocketFrameType.BINARY) {
+                    onBinary(webSocketFrame.array(), last);
+                } else {
+                    throw new IllegalStateException("Unexpected frame type: " + lastFrameType + " (expected: " +
+                                                    WebSocketFrameType.TEXT + " or " +
+                                                    WebSocketFrameType.BINARY + ')');
+                }
+                break;
+            case CLOSE:
+                final CloseWebSocketFrame closeWebSocketFrame = (CloseWebSocketFrame) webSocketFrame;
+                listener.onClose(this, closeWebSocketFrame.status().code(), closeWebSocketFrame.reasonPhrase());
+                break;
+            case PING:
+            case PONG:
+                request();
+                break;
+        }
+    }
+
+    private void onText(String text, boolean last) {
+        textBuffer.append(text);
+        if (last) {
+            final String value = textBuffer.toString();
+            textBuffer.setLength(0);
+            listener.onMessage(this, value);
+        } else {
+            request();
+        }
+    }
+
+    private void onBinary(byte[] bytes, boolean last) {
+        binaryBuffer.add(bytes);
+        if (last) {
+            int size = 0;
+            for (byte[] b : binaryBuffer) {
+                size += b.length;
+            }
+            final ByteBuffer buffer = ByteBuffer.allocate(size);
+            for (;;) {
+                final byte[] binary = binaryBuffer.poll();
+                if (binary == null) {
+                    break;
+                } else {
+                    buffer.put(binary);
+                }
+            }
+            buffer.flip();
+            listener.onMessage(this, buffer);
+        } else {
+            request();
+        }
```

Comment:
Question) taking a quick look at `WebSocketFrameDecoder`, it seems like a single frame corresponds to a single message.

Is there no need to request() also for a continuation frame?

```suggestion
        request();
```

Reply:
> Is there no need to request() also for a continuation frame?

`WatcherWebSocketListener` calls `request()` whenever it receives a new message.
So I called `request()` here only for a continuation frame?
https://github.com/fabric8io/kubernetes-client/blob/56a6c2c4f336cc6f64c19029a55c2d3d0289344f/kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/internal/WatcherWebSocketListener.java#L54-L54

Let me add a comment for that.

Reply:
I see, understood 👍     I assume this implementation will change when https://github.com/line/armeria/pull/5357 is merged so I'll take another look then